Added bisection method for 3D lines

Jaime02 opened the pull request:
This PR adds a new method for generating the bisection line of two 3D lines. It currently needs a code quality revision and a better documentation.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Added method for getting the bisectors of two lines

#### Other comments

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* geometry
  * Added a new method for calculating the bisectors of two lines.

<!-- END RELEASE NOTES -->

Jaime02 committed:
Added bisection method for 3D lines

This PR adds a new method for generating the bisection line of two 3D lines. It currently needs tests, a code quality revision and an implementation for more dimensions.

Jaime02 committed:
Fixed whitespace

Jaime02 commented:
I think that the current algorithm is wrong.

Jaime02 committed:
Fixed wrong variable name

Jaime02 commented:
The tests have failed because unpacking operator (*) is no supported in python 2.7

oscarbenjamin commented:
Python 2.7 is not supported any more. The tests on Travis just check to see that importing sympy gives an importerror but there is a problem with it. I have the fix for .travis.yml in #18244. Anyone is free to make a PR with that.

Jaime02 commented:
So... what should I do? Waiting until the test drop python 2 support?

oscarbenjamin commented:
Closed and reopened to restart the tests merged with current master. Python 3 only syntax should be possible now after #18244

Jaime02 committed:
Added bisection test

Jaime02 committed:
Fixed variable name at docstring

Jaime02 commented:
Wait the algorithm does not work with direction ratios which are greater than 90º

Jaime02 committed:
Merge branch 'master' into master

Jaime02 committed:
Too more newlines at the end

smichr commented on sympy/geometry/line.py:
```diff
@@ -2526,6 +2526,44 @@ def equation(self, x='x', y='y', z='z', k=None):
         return Tuple(*[i.subs(k, kk).as_numer_denom()[0] for i in eqs])
 
+    def bisection(self, line):
+        """Returns the line which is the bisection between the self line and another one.
```
I would suggest the name "bisector" and say that it is the "line which bisects the angle formed at the intersection of the two lines and lying in the same plane"

But I wonder if, instead of raising an error for non-intersecting lines, a meaningful line could be returned which passes perpendicularly through the midpoint of the shortest segment connecting the two lines.
```python
>>> r1
Line3D(Point3D(0, 0, 0), Point3D(1, 1, 1))
>>> r4
Line3D(Point3D(1, 1, 0), Point3D(0, 2, 1))
>>> r1.intersection(r4)
[]
>>> t=Dummy()
>>> p1=r1.arbitrary_point(t)
>>> p4=r4.arbitrary_point(t)
>>> tmin=solve(p1.distance(p4).diff(t))[0] # find t which minimizes distance between arb pts
>>> m=Segment(p1,p4).midpoint.subs(t,tmin) # find mdpoint of shortest segment connecting lines
>>> d = Matrix(r1.direction_ratio).cross(Matrix(r4.direction_ratio)) # new direction
>>> Line(m,direction_ratio=d) # "bisector" line
Line3D(Point3D(1/2, 1, 1/2), Point3D(1/2, -1, 5/2))
```
So the definition of "bisector" for two lines is the line passing through either the intersection of the two lines or the line passing perpendicularly through the midpoint of the shortest segment between the two lines.

Jaime02 replied:
Can you change the current code? Also the current code does not work with direction ratios greater than 90º and I don't know how to solve it.

Jaime02 replied:
That is a good addition, but you are talking about a line perpendicular to a segment at certain point contained in the segment. There are infinite lines which satisfy that condition which define a plane.

Jaime02 commented:
https://stackoverflow.com/questions/59809904/find-the-bisection-of-two-3d-lines

Jaime02 committed:
Changed the algorithm

References:
https://stackoverflow.com/questions/59809904/find-the-bisection-of-two-3d-lines

Not finished at all. Tests are pending.

Jaime02 committed:
imported intersection function

I dont know if this will work

Jaime02 committed:
Changed method's name 

bisections -> bisector
Use Intersection instead of intersection

oscarbenjamin commented on sympy/geometry/tests/test_line.py:
```diff
@@ -749,6 +749,16 @@ def test_parameter_value():
+
+    assert r1.intersection(r2) == [Point3D(1, 1, 1)]
+
+    r3 = r1.bisection(r2)
```
The method you added is called `bisector`

Jaime02 replied:
Fixed, thanks!

Jaime02 committed:
Fixed method's name at test

Jaime02 commented:
How can I use the geometry method called `intersection` located in` Lib\site-packages\sympy\geometry\util.py ?`

smichr commented:
> How can I use the geometry method called `intersection`

```
>>> help(intersection)
Help on function intersection in module sympy.geometry.util:

intersection(*entities, **kwargs)
    The intersection of a collection of GeometryEntity instances.
    
    Parameters
    ==========
...
```

smichr commented on sympy/geometry/line.py:
```diff
@@ -2526,6 +2527,49 @@ def equation(self, x='x', y='y', z='z', k=None):
+            vector = [i/length for i in vector]
+            return vector
+
+        d1 = normalize(self.direction_ratio)
```
Note:
```python
>>> l=Line((1,2,3),(3,6,8))
>>> l.direction_ratio
[2, 4, 5]
>>> Point(l.direction_ratio).unit.args
(2*sqrt(5)/15, 4*sqrt(5)/15, sqrt(5)/3)
```

Jaime02 replied:
> > How can I use the geometry method called `intersection`
> 
> ```
> >>> help(intersection)
> Help on function intersection in module sympy.geometry.util:
> 
> intersection(*entities, **kwargs)
>     The intersection of a collection of GeometryEntity instances.
>     
>     Parameters
>     ==========
> ...
> ```

I mean, how can I use that function inside the bisection method? If I put a `from sympy.geometry import intersection` it shows an `ImportError`

Jaime02 replied:
What you mean is that the direction ratio when converted to a point gets normalized?

smichr replied:
`d1 = Point(self.direction_ratio).unit.args` can be used instead of creating the normalize function.

smichr commented on sympy/geometry/line.py:
```diff
@@ -2526,6 +2527,49 @@ def equation(self, x='x', y='y', z='z', k=None):
                 break
         return Tuple(*[i.subs(k, kk).as_numer_denom()[0] for i in eqs])
 
+    def bisector(self, line):
```
I see that this will return two lines so plural should be used. But perhaps it is better called `loci` and the more general definition I gave could be used even when the lines don't intersect at a point. Bisector should probably be reserved for angles and segments.

Jaime02 replied:
What you mean by "loc"i is [this](https://en.wikipedia.org/wiki/Locus_(mathematics))?
You are right, it should be plural. However I still do not know how to implement your suggestion of two lines which do not intersect.

smichr replied:
See [here](https://github.com/sympy/sympy/pull/18358#discussion_r368295351): `line(m, direction_ratio=d)` is one line. The other would be rotated 90 degrees and perpendicular to the same segment, I think.

smichr replied:
> What you mean by "loc"i is [this](https://en.wikipedia.org/wiki/Locus_(mathematics))?

yes

Jaime02 replied:
I think I am not able to code that functonality, my math is quite limited. Could you do it, please?

Jaime02 commented:
> > How can I use the geometry method called `intersection`
> 
> ```
> >>> help(intersection)
> Help on function intersection in module sympy.geometry.util:
> 
> intersection(*entities, **kwargs)
>     The intersection of a collection of GeometryEntity instances.
>     
>     Parameters
>     ==========
> ...
> ```
I mean, how can I use that function inside the bisection method? If I put a `from sympy.geometry import intersection` it shows an `ImportError`

smichr commented:
> If I put a `from sympy.geometry import`

try `from sympy.geometry.util import intersection` (see the path in the first line of help)

Jaime02 committed:
Fixed intersection function

And renamed method to be in plural because it returns two lines when the lines have an intersection point

Jaime02 committed:
Updated method's name

Jaime02 commented:
Why do I get this strange fraction result? 
`from sympy import Point3D, Line3D, intersection

r1 = Line3D(Point3D(0, 0, 0), Point3D(1, 1, 1))
r2 = Line3D(Point3D(2, 0, 0), Point3D(-1, 1, 1))

print(intersection(r1, r2))
print(r1.bisectors(r2))`

Output:
`[Point3D(1/2, 1/2, 1/2)]
[Line3D(Point3D(1/2, 1/2, 1/2), Point3D(34563247091267/200000000000000, 137886161376739/100000000000000, 137886161376739/100000000000000)), Line3D(Point3D(1/2, 1/2, 1/2), Point3D(49547107573073/25000000000000, 387919462305931/500000000000000, 387919462305931/500000000000000))]`

czgdp1807 commented:
@Jaime02 Are you still working on it?

Jaime02 commented:
@czgdp1807 I left the code as it is right now. Tests are incomplete and I have found many errors which I do not know how to fix.

Jaime02 committed:
Changed test

Jaime02 commented:
How can I normalize a vector properly?

smichr commented:
> How can I normalize a vector properly?

if you have the 3 components of the vector and pass them as coordinates for Point, the `unit` property will given their values so the length of the vector is 1:

```python
>>> Point(2,3,4).unit
Point3D(2*sqrt(29)/29, 3*sqrt(29)/29, 4*sqrt(29)/29)
>>> _.args
(2*sqrt(29)/29, 3*sqrt(29)/29, 4*sqrt(29)/29)
>>> sqrt(sum(i**2 for i in _))
1
```

smichr commented:
> Why do I get this strange fraction result?

Looks like conversion of Floats to Rationals. Somewhere the Rationals were converted to Floats during the calculation.

Jaime02 committed:
Changed normalizing function

There is a built in way of normalizating

Jaime02 committed:
Added 2D line bisection

The algorithm has been modified a bit to work in two dimensions

Jaime02 commented:
Shall the `from __future__ import division, print_function` statement stay there?

Jaime02 committed:
Fixed copypaste error

Jaime02 committed:
Fixed missing imports

Jaime02 committed:
Finally fixed imports

Jaime02 committed:
FINALLY FIXED IMPORTS

Jaime02 committed:
FiNaLlY fIxEd ImPoRtS

Jaime02 commented:
How can I use the own class in the doctests? They fail because of that

asmeurer commented:
You need to add lines to the doctest that import the class.

Jaime02 committed:
Fixed missing imports at doctest

smichr commented on sympy/geometry/line.py:
```diff
@@ -2526,6 +2578,50 @@ def equation(self, x='x', y='y', z='z', k=None):
+        d1 = normalize(self.direction_ratio)
+        d2 = normalize(line.direction_ratio)
+
+        bis1 = Line3D(point[0], direction_ratio=[d1[i] + d2[i] for i in range(3)])
```
Seeing what you are doing with d1 and d2 suggest this:
```python
d1 = Point(self.direction_ratio).unit
d2 = Point(line.direction_ratio).unit
bis1 = Line3D(point[0], direction_ratio=d1 + d2)
bis2 = Line3D(point[0], direction_ratio=d1 - d2)
```

Jaime02 replied:
Yes that's a clever optimization! Thanks!
If it is known that the points are 2D, should it use Point2D instead of Point? 
The same applies for the 3D case.

smichr replied:
SymPy does the right thing for a given number of arguments:
```python
>>> Point(2,3,4)
Point3D(2, 3, 4)
>>> Line((1,2,3),(3,2,3))
Line3D(Point3D(1, 2, 3), Point3D(3, 2, 3))
```
For this reason, too, it should only be necessary to define this function once and perhaps call it from Line2D and Line3D; in the function you could use Point and Line instead of the 3D or 2D decorated forms.

Jaime02 replied:
Yes but explicit is better than implicit, right?

smichr replied:
See the LinearEntity class for examples of methods that are written without reference to dimension but apply to 2D and 3D. Look, too, for use of `ambient_dimension` to make sure that the case of a 2D line intersecting a 3D line is considered and handled properly.

smichr commented on sympy/geometry/line.py:
```diff
@@ -201,7 +202,7 @@ def angle_between(l1, l2):
 
         v1, v2 = l1.direction, l2.direction
-        return acos(v1.dot(v2)/(abs(v1)*abs(v2)))
+        return acos(v1.dot(v2) / (abs(v1) * abs(v2)))
```
These changes can be removed. We use space around `+-` but not `*/` but this is loosely enforced. If you do 
```
git reset head~22 # or however many commits you have when you get to this
git add -p # then say y/n to each change as prompted, saying n to all these whitespace changes
git commit -m # add bisection method for Line' (or whatever title you want
git reset --hard # to throw away other changes
git push -f Jaime02 master
```

that should update things here under a single commit.

Jaime02 committed:
Simplified normalization and undone whitespaces

Jaime02 committed:
Some / had been changed to *

Pycharm's fault

Jaime02 committed:
direction_ratio instead of direction in a 2D line

Jaime02 committed:
Changed Line2D to Line

Jaime02 committed:
Removed unused import

Jaime02 committed:
It seems like Point3D import is necessary

smichr committed:
move bisectors to LinearEntity

smichr commented:
I wonder if there is still a better name. What this is doing is creating a set of axes in the plane of the two linear entities which passes through their intersection. I also sent a [PR](https://github.com/Jaime02/sympy/pull/1) which moves code to LinearEntity and allows for any LinearEntities to be used (see tests).

It would be good to add tests for cases which are to raise (not return) errors, too.

Jaime02 committed:
Merge pull request #1 from smichr/linbis

moved bisectors to LinearEntity

Jaime02 commented:
> I wonder if there is still a better name.
I'm not english native, if you found a better name feel free to change it. [Maybe this can help 
](https://en.wikipedia.org/wiki/Bisection)

> I also sent a PR which moves code to LinearEntity
That is a good PR, thank you so much! I will add more tests when I had time.

smichr commented:
When will this method be useful?

Jaime02 commented:
> When will this method be useful?

That's a good question. I personally use sympy because it has very good geometry entities. My technical drawing program needed a tool to get the bisections of two lines and it was not implemented in Sympy. [After asking for help](https://stackoverflow.com/questions/59809904/find-the-bisection-of-two-3d-lines) I coded it, so I thought it would be interesting to have this feature built-in in Sympy, I think many users who use geometry will found this feature useful.

For example, the blue lines are the bisections of the two black lines.
![image](https://user-images.githubusercontent.com/38530589/75703711-4899d380-5cb8-11ea-9265-beb97e08e24e.png)

smichr commented:
> it would be good to get some feedback [about this](https://github.com/sympy/sympy/pull/18358#issuecomment-593208545)

oscarbenjamin commented:
I'm not sure I understand what this does.

Given two non-parallel lines in 3D space that intersect the two lines are contained within a plane. The method here returns two other lines that are in that plane and orthogonal to each other. Is that correct?

The plane contains many pairs of orthogonal lines. Presumably these two intersect where the original lines intersect. What determines their orientation?

Jaime02 commented:
> I'm not sure I understand what this does.
> 
> Given two non-parallel lines in 3D space that intersect the two lines are contained within a plane. The method here returns two other lines that are in that plane and orthogonal to each other. Is that correct?
> 
> The plane contains many pairs of orthogonal lines. Presumably these two intersect where the original lines intersect. What determines their orientation?

The most important property of the returned lines is that any point in that lines is equidistant to the two initial lines. It has other geometrical properties.

Commit:
Revert "Initial commit i think. (I dont know how git works at all)"

This reverts commit 1c80758b5b3ba6bd5876c007587f68e1dcd26d0c.

Commit:
Initial commit i think. (I dont know how git works at all)

Jaime02 commented:
Oops, I think I commited the lastest sympy master source code into my current PR, I don't know how to use git properly.

smichr commented:
> What determines their orientation?

This returns the pair of lines that are in the same plane as the original two, orthogonal to each other and bisecting the angles of the originals. So if the x and y axes were passed, lines and +/-45 deg would be returned.

smichr commented:
Can you start fresh from a new master and apply your diff? Here is the current diff relative to master:
```diff
diff --git a/sympy/geometry/line.py b/sympy/geometry/line.py
index 12c162c..cd52e1a 100644
--- a/sympy/geometry/line.py
+++ b/sympy/geometry/line.py
@@ -18,28 +18,28 @@
 """
 from __future__ import division, print_function
 
-
 from sympy import Expr
 from sympy.core import S, sympify
 from sympy.core.compatibility import ordered
+from sympy.core.containers import Tuple
+from sympy.core.decorators import deprecated
 from sympy.core.numbers import Rational, oo
 from sympy.core.relational import Eq
 from sympy.core.symbol import _symbol, Dummy
-from sympy.functions.elementary.trigonometric import (_pi_coeff as pi_coeff, acos, tan, atan2)
 from sympy.functions.elementary.piecewise import Piecewise
-from sympy.logic.boolalg import And
-from sympy.simplify.simplify import simplify
+from sympy.functions.elementary.trigonometric import (_pi_coeff as pi_coeff, acos, tan, atan2)
 from sympy.geometry.exceptions import GeometryError
-from sympy.core.containers import Tuple
-from sympy.core.decorators import deprecated
-from sympy.sets import Intersection
+from sympy.geometry.util import intersection
+from sympy.logic.boolalg import And
 from sympy.matrices import Matrix
+from sympy.sets import Intersection
+from sympy.simplify.simplify import simplify
 from sympy.solvers.solveset import linear_coeffs
-from .entity import GeometryEntity, GeometrySet
-from .point import Point, Point3D
+from sympy.utilities.exceptions import SymPyDeprecationWarning
 from sympy.utilities.misc import Undecidable, filldedent
 
-from sympy.utilities.exceptions import SymPyDeprecationWarning
+from .entity import GeometryEntity, GeometrySet
+from .point import Point, Point3D
 
 
 class LinearEntity(GeometrySet):
@@ -96,7 +96,6 @@ def _span_test(self, other):
         A point x is 'in front' of a point y if x.dot(y) >= 0.  Return
         -1 if `other` is behind `self.p1`, 0 if `other` is `self.p1` and
         and 1 if `other` is in front of `self.p1`."""
-
         if self.p1 == other:
             return 0
 
@@ -341,7 +340,6 @@ def are_concurrent(*lines):
         False
 
         """
-
         common_points = Intersection(*lines)
         if common_points.is_FiniteSet and len(common_points) == 1:
             return True
@@ -456,9 +454,9 @@ def intersect_parallel_ray_and_segment(ray, seg):
                 return []
             elif st1 >= 0 and st2 >= 0:
                 return [seg]
-            elif st1 >= 0: # st2 < 0:
+            elif st1 >= 0:  # st2 < 0:
                 return [Segment(ray.p1, seg.p1)]
-            elif st2 >= 0: # st1 < 0:
+            elif st2 >= 0:  # st1 < 0:
                 return [Segment(ray.p1, seg.p2)]
 
         def intersect_parallel_segments(seg1, seg2):
@@ -534,7 +532,7 @@ def intersect_parallel_segments(seg1, seg2):
                     return [line_intersection]
 
                 if ((isinstance(self, Line) or
-                        self.contains(line_intersection)) and
+                     self.contains(line_intersection)) and
                         other.contains(line_intersection)):
                     return [line_intersection]
                 return []
@@ -589,7 +587,6 @@ def is_parallel(l1, l2):
         False
 
         """
-
         if not isinstance(l1, LinearEntity) and not isinstance(l2, LinearEntity):
             raise TypeError('Must pass only LinearEntity objects')
 
@@ -951,7 +948,6 @@ def projection(self, other):
         Segment3D(Point3D(10/3, 10/3, 13/3), Point3D(5, 5, 6))
 
         """
-
         if not isinstance(other, GeometryEntity):
             other = Point(other, dim=self.ambient_dimension)
 
@@ -1029,6 +1025,64 @@ def random_point(self, seed=None):
             raise NotImplementedError('unhandled line type')
         return pt.subs(t, Rational(v))
 
+    def bisectors(self, other):
+        """Returns the perpendicular lines which pass through the intersections
+        of self and other that are in the same plane.
+
+        Parameters
+        ==========
+
+        line : Line3D
+
+        Returns
+        =======
+
+        list: two Line instances
+
+        Examples
+        ========
+
+        >>> from sympy.geometry import Point3D, Line3D
+        >>> r1 = Line3D(Point3D(0, 0, 0), Point3D(1, 0, 0))
+        >>> r2 = Line3D(Point3D(0, 0, 0), Point3D(0, 1, 0))
+        >>> r1.bisectors(r2)
+        [Line3D(Point3D(0, 0, 0), Point3D(1, 1, 0)), Line3D(Point3D(0, 0, 0), Point3D(1, -1, 0))]
+
+        """
+        if not isinstance(other, LinearEntity):
+            raise GeometryError("Expecting LinearEntity, not %s" % other)
+
+        l1, l2 = self, other
+
+        # make sure dimensions match or else a warning will rise from
+        # intersection calculation
+        if l1.p1.ambient_dimension != l2.p1.ambient_dimension:
+            if isinstance(l1, Line2D):
+                l1, l2 = l2, l1
+            _, p1 = Point._normalize_dimension(l1.p1, l2.p1, on_morph='ignore')
+            _, p2 = Point._normalize_dimension(l1.p2, l2.p2, on_morph='ignore')
+            l2 = Line(p1, p2)
+
+        point = intersection(l1, l2)
+
+        # Three cases: Lines may intersect in a point, may be equal or may not intersect.
+        if not point:
+            raise GeometryError("The lines do not intersect")
+        else:
+            pt = point[0]
+            if isinstance(pt, Line):
+                # Intersection is a line because both lines are coincident
+                return [self]
+
+
+        d1 = l1.direction.unit
+        d2 = l2.direction.unit
+
+        bis1 = Line(pt, pt + d1 + d2)
+        bis2 = Line(pt, pt + d1 - d2)
+
+        return [bis1, bis2]
+
 
 class Line(LinearEntity):
     """An infinite line in space.
@@ -1100,7 +1154,6 @@ class Line(LinearEntity):
     >>> Line(Eq(3*a + b, -18), x='a', y=b)
     Line2D(Point2D(0, -18), Point2D(1, -21))
     """
-
     def __new__(cls, *args, **kwargs):
         from sympy.geometry.util import find
 
@@ -1128,7 +1181,7 @@ def __new__(cls, *args, **kwargs):
                 if len(args) > 1:
                     p2 = args[1]
                 else:
-                    p2=None
+                    p2 = None
 
                 if isinstance(p1, LinearEntity):
                     if p2:
@@ -1332,7 +1385,6 @@ def _svg(self, scale_factor=1., fill_color="#66cc99"):
         fill_color : str, optional
             Hex string for fill color. Default is "#66cc99".
         """
-
         from sympy.core.evalf import N
 
         verts = (N(self.p1), N(self.p2))
@@ -1343,7 +1395,7 @@ def _svg(self, scale_factor=1., fill_color="#66cc99"):
             '<path fill-rule="evenodd" fill="{2}" stroke="#555555" '
             'stroke-width="{0}" opacity="0.6" d="{1}" '
             'marker-start="url(#markerCircle)" marker-end="url(#markerArrow)"/>'
-            ).format(2. * scale_factor, path, fill_color)
+        ).format(2.*scale_factor, path, fill_color)
 
     def contains(self, other):
         """
@@ -1580,10 +1632,9 @@ def contains(self, other):
         >>> s2 = Segment3D(p2, p1)
         >>> s.contains(s2)
         True
-        >>> s.contains((p1 + p2) / 2)
+        >>> s.contains((p1 + p2)/2)
         True
         """
-
         if not isinstance(other, GeometryEntity):
             other = Point(other, dim=self.ambient_dimension)
         if isinstance(other, Point):
@@ -1820,7 +1871,6 @@ def bounds(self):
         rectangle for the geometric figure.
 
         """
-
         verts = self.points
         xs = [p.x for p in verts]
         ys = [p.y for p in verts]
@@ -1985,7 +2035,6 @@ def _svg(self, scale_factor=1., fill_color="#66cc99"):
         fill_color : str, optional
             Hex string for fill color. Default is "#66cc99".
         """
-
         from sympy.core.evalf import N
 
         verts = (N(self.p1), N(self.p2))
@@ -1996,7 +2045,7 @@ def _svg(self, scale_factor=1., fill_color="#66cc99"):
             '<path fill-rule="evenodd" fill="{2}" stroke="#555555" '
             'stroke-width="{0}" opacity="0.6" d="{1}" '
             'marker-start="url(#markerReverseArrow)" marker-end="url(#markerArrow)"/>'
-            ).format(2. * scale_factor, path, fill_color)
+        ).format(2.*scale_factor, path, fill_color)
 
     @property
     def coefficients(self):
@@ -2338,7 +2387,6 @@ def _svg(self, scale_factor=1., fill_color="#66cc99"):
         fill_color : str, optional
             Hex string for fill color. Default is "#66cc99".
         """
-
         from sympy.core.evalf import N
 
         verts = (N(self.p1), N(self.p2))
@@ -2347,7 +2395,7 @@ def _svg(self, scale_factor=1., fill_color="#66cc99"):
         return (
             '<path fill-rule="evenodd" fill="{2}" stroke="#555555" '
             'stroke-width="{0}" opacity="0.6" d="{1}" />'
-            ).format(2. * scale_factor, path, fill_color)
+        ).format(2.*scale_factor, path, fill_color)
 
 
 class LinearEntity3D(LinearEntity):
@@ -2368,7 +2416,6 @@ class LinearEntity3D(LinearEntity):
 
     This is a base class and is not meant to be instantiated.
     """
-
     def __new__(cls, p1, p2, **kwargs):
         p1 = Point3D(p1, dim=3)
         p2 = Point3D(p2, dim=3)
@@ -2457,7 +2504,6 @@ class Line3D(LinearEntity3D, Line):
     >>> L.points
     (Point3D(2, 3, 4), Point3D(3, 5, 1))
     """
-
     def __new__(cls, p1, pt=None, direction_ratio=[], **kwargs):
         if isinstance(p1, LinearEntity3D):
             if pt is not None:
@@ -2509,9 +2555,9 @@ def equation(self, x='x', y='y', z='z', k=None):
         """
         if k is not None:
             SymPyDeprecationWarning(
-                            feature="equation() no longer needs 'k'",
-                            issue=13742,
-                            deprecated_since_version="1.2").warn()
+                feature="equation() no longer needs 'k'",
+                issue=13742,
+                deprecated_since_version="1.2").warn()
         from sympy import solve
         x, y, z, k = [_symbol(i, real=True) for i in (x, y, z, 'k')]
         p1, p2 = self.points
@@ -2574,7 +2620,6 @@ class Ray3D(LinearEntity3D, Ray):
     [1, 2, -4]
 
     """
-
     def __new__(cls, p1, pt=None, direction_ratio=[], **kwargs):
         from sympy.utilities.misc import filldedent
         if isinstance(p1, LinearEntity3D):
@@ -2731,7 +2776,6 @@ class Segment3D(LinearEntity3D, Segment):
     Point3D(5/2, 2, 8)
 
     """
-
     def __new__(cls, p1, p2, **kwargs):
         p1 = Point(p1, dim=3)
         p2 = Point(p2, dim=3)
diff --git a/sympy/geometry/tests/test_line.py b/sympy/geometry/tests/test_line.py
index 30f9a7e..b343565 100644
--- a/sympy/geometry/tests/test_line.py
+++ b/sympy/geometry/tests/test_line.py
@@ -752,6 +752,20 @@ def test_parameter_value():
     raises(ValueError, lambda: l.parameter_value((0, 0), t))
 
 
+def test_bisectors():
+    r1 = Line3D(Point3D(0, 0, 0), Point3D(1, 0, 0))
+    r2 = Line3D(Point3D(0, 0, 0), Point3D(0, 1, 0))
+    bisections = r1.bisectors(r2)
+    assert bisections == [Line3D(Point3D(0, 0, 0), Point3D(1, 1, 0)),
+        Line3D(Point3D(0, 0, 0), Point3D(1, -1, 0))]
+    ans = [Line3D(Point3D(0, 0, 0), Point3D(1, 0, 1)),
+        Line3D(Point3D(0, 0, 0), Point3D(-1, 0, 1))]
+    l1 = (0, 0, 0), (0, 0, 1)
+    l2 = (0, 0), (1, 0)
+    for a, b in cartes((Line, Segment, Ray), repeat=2):
+        assert a(*l1).bisectors(b(*l2)) == ans
+
+
 def test_issue_8615():
     a = Line3D(Point3D(6, 5, 0), Point3D(6, -6, 0))
     b = Line3D(Point3D(6, -1, 19/10), Point3D(6, -1, 0))
```

Jaime02 commented:
> > What determines their orientation?
> 
> This returns the pair of lines that are in the same plane as the original two, orthogonal to each other and bisecting the angles of the originals. So if the x and y axes were passed, lines and +/-45 deg would be returned.

That's it!

Jaime02 commented:
> Can you start fresh from a new master and apply your diff? 

Shall I close this PR and open a new one?

smichr commented:
> Shall I close this PR and open a new one?

No need. As long as this one is open when you push to it, it will be updated. You will likely need the `git push -f` version.

Jaime02 commented:
I don't know what to do with the current master

asmeurer commented:
@Jaime02 one issue is that you committed these changes to your master branch. You should avoid doing that. Instead, you should always create a new branch (`git checkout -b branch-name`) before committing anything. That way, your `master` branch stays the same. 

It is possible to fix it, but it may be simpler for you to reclone the sympy repository (from sympy/sympy, not your fork), and create a new branch there, then create a new PR from that, using the diff noted by @smichr. After you do that, you will also want to revert your master branch to the original master branch, which you can do with `git push -f`.

Jaime02 commented:
I don't know whether I have done it properly, I cloned the lastest sympy master so there are many changes

smichr commented:
* Apply the diff described [here](https://github.com/sympy/sympy/pull/18358#issuecomment-599038744) to your fresh master, 
* commit changes
* `git push -f Jaime02 master` to update this PR
* `git branch line-bisect` to create a `line-bisect` branch in case you need it again.
* then revert the changes to your master with `git reset head~1`

At that point you will have the standard master, a line-bisect branch, and the PR will be up to date.

And then you will never create another PR using master ;-)

Jaime02 commented:
> Can you start fresh from a new master and apply your diff? Here is the current diff relative to master:
> 
> ```diff
> diff --git a/sympy/geometry/line.py b/sympy/geometry/line.py
> index 12c162c..cd52e1a 100644
> --- a/sympy/geometry/line.py
> +++ b/sympy/geometry/line.py
> @@ -18,28 +18,28 @@
>  """
>  from __future__ import division, print_function
>  
> -
>  from sympy import Expr
>  from sympy.core import S, sympify
>  from sympy.core.compatibility import ordered
> +from sympy.core.containers import Tuple
> +from sympy.core.decorators import deprecated
>  from sympy.core.numbers import Rational, oo
>  from sympy.core.relational import Eq
>  from sympy.core.symbol import _symbol, Dummy
> -from sympy.functions.elementary.trigonometric import (_pi_coeff as pi_coeff, acos, tan, atan2)
>  from sympy.functions.elementary.piecewise import Piecewise
> -from sympy.logic.boolalg import And
> -from sympy.simplify.simplify import simplify
> +from sympy.functions.elementary.trigonometric import (_pi_coeff as pi_coeff, acos, tan, atan2)
>  from sympy.geometry.exceptions import GeometryError
> -from sympy.core.containers import Tuple
> -from sympy.core.decorators import deprecated
> -from sympy.sets import Intersection
> +from sympy.geometry.util import intersection
> +from sympy.logic.boolalg import And
>  from sympy.matrices import Matrix
> +from sympy.sets import Intersection
> +from sympy.simplify.simplify import simplify
>  from sympy.solvers.solveset import linear_coeffs
> -from .entity import GeometryEntity, GeometrySet
> -from .point import Point, Point3D
> +from sympy.utilities.exceptions import SymPyDeprecationWarning
>  from sympy.utilities.misc import Undecidable, filldedent
>  
> -from sympy.utilities.exceptions import SymPyDeprecationWarning
> +from .entity import GeometryEntity, GeometrySet
> +from .point import Point, Point3D
>  
>  
>  class LinearEntity(GeometrySet):
> @@ -96,7 +96,6 @@ def _span_test(self, other):
>          A point x is 'in front' of a point y if x.dot(y) >= 0.  Return
>          -1 if `other` is behind `self.p1`, 0 if `other` is `self.p1` and
>          and 1 if `other` is in front of `self.p1`."""
> -
>          if self.p1 == other:
>              return 0
>  
> @@ -341,7 +340,6 @@ def are_concurrent(*lines):
>          False
>  
>          """
> -
>          common_points = Intersection(*lines)
>          if common_points.is_FiniteSet and len(common_points) == 1:
>              return True
> @@ -456,9 +454,9 @@ def intersect_parallel_ray_and_segment(ray, seg):
>                  return []
>              elif st1 >= 0 and st2 >= 0:
>                  return [seg]
> -            elif st1 >= 0: # st2 < 0:
> +            elif st1 >= 0:  # st2 < 0:
>                  return [Segment(ray.p1, seg.p1)]
> -            elif st2 >= 0: # st1 < 0:
> +            elif st2 >= 0:  # st1 < 0:
>                  return [Segment(ray.p1, seg.p2)]
>  
>          def intersect_parallel_segments(seg1, seg2):
> @@ -534,7 +532,7 @@ def intersect_parallel_segments(seg1, seg2):
>                      return [line_intersection]
>  
>                  if ((isinstance(self, Line) or
> -                        self.contains(line_intersection)) and
> +                     self.contains(line_intersection)) and
>                          other.contains(line_intersection)):
>                      return [line_intersection]
>                  return []
> @@ -589,7 +587,6 @@ def is_parallel(l1, l2):
>          False
>  
>          """
> -
>          if not isinstance(l1, LinearEntity) and not isinstance(l2, LinearEntity):
>              raise TypeError('Must pass only LinearEntity objects')
>  
> @@ -951,7 +948,6 @@ def projection(self, other):
>          Segment3D(Point3D(10/3, 10/3, 13/3), Point3D(5, 5, 6))
>  
>          """
> -
>          if not isinstance(other, GeometryEntity):
>              other = Point(other, dim=self.ambient_dimension)
>  
> @@ -1029,6 +1025,64 @@ def random_point(self, seed=None):
>              raise NotImplementedError('unhandled line type')
>          return pt.subs(t, Rational(v))
>  
> +    def bisectors(self, other):
> +        """Returns the perpendicular lines which pass through the intersections
> +        of self and other that are in the same plane.
> +
> +        Parameters
> +        ==========
> +
> +        line : Line3D
> +
> +        Returns
> +        =======
> +
> +        list: two Line instances
> +
> +        Examples
> +        ========
> +
> +        >>> from sympy.geometry import Point3D, Line3D
> +        >>> r1 = Line3D(Point3D(0, 0, 0), Point3D(1, 0, 0))
> +        >>> r2 = Line3D(Point3D(0, 0, 0), Point3D(0, 1, 0))
> +        >>> r1.bisectors(r2)
> +        [Line3D(Point3D(0, 0, 0), Point3D(1, 1, 0)), Line3D(Point3D(0, 0, 0), Point3D(1, -1, 0))]
> +
> +        """
> +        if not isinstance(other, LinearEntity):
> +            raise GeometryError("Expecting LinearEntity, not %s" % other)
> +
> +        l1, l2 = self, other
> +
> +        # make sure dimensions match or else a warning will rise from
> +        # intersection calculation
> +        if l1.p1.ambient_dimension != l2.p1.ambient_dimension:
> +            if isinstance(l1, Line2D):
> +                l1, l2 = l2, l1
> +            _, p1 = Point._normalize_dimension(l1.p1, l2.p1, on_morph='ignore')
> +            _, p2 = Point._normalize_dimension(l1.p2, l2.p2, on_morph='ignore')
> +            l2 = Line(p1, p2)
> +
> +        point = intersection(l1, l2)
> +
> +        # Three cases: Lines may intersect in a point, may be equal or may not intersect.
> +        if not point:
> +            raise GeometryError("The lines do not intersect")
> +        else:
> +            pt = point[0]
> +            if isinstance(pt, Line):
> +                # Intersection is a line because both lines are coincident
> +                return [self]
> +
> +
> +        d1 = l1.direction.unit
> +        d2 = l2.direction.unit
> +
> +        bis1 = Line(pt, pt + d1 + d2)
> +        bis2 = Line(pt, pt + d1 - d2)
> +
> +        return [bis1, bis2]
> +
>  
>  class Line(LinearEntity):
>      """An infinite line in space.
> @@ -1100,7 +1154,6 @@ class Line(LinearEntity):
>      >>> Line(Eq(3*a + b, -18), x='a', y=b)
>      Line2D(Point2D(0, -18), Point2D(1, -21))
>      """
> -
>      def __new__(cls, *args, **kwargs):
>          from sympy.geometry.util import find
>  
> @@ -1128,7 +1181,7 @@ def __new__(cls, *args, **kwargs):
>                  if len(args) > 1:
>                      p2 = args[1]
>                  else:
> -                    p2=None
> +                    p2 = None
>  
>                  if isinstance(p1, LinearEntity):
>                      if p2:
> @@ -1332,7 +1385,6 @@ def _svg(self, scale_factor=1., fill_color="#66cc99"):
>          fill_color : str, optional
>              Hex string for fill color. Default is "#66cc99".
>          """
> -
>          from sympy.core.evalf import N
>  
>          verts = (N(self.p1), N(self.p2))
> @@ -1343,7 +1395,7 @@ def _svg(self, scale_factor=1., fill_color="#66cc99"):
>              '<path fill-rule="evenodd" fill="{2}" stroke="#555555" '
>              'stroke-width="{0}" opacity="0.6" d="{1}" '
>              'marker-start="url(#markerCircle)" marker-end="url(#markerArrow)"/>'
> -            ).format(2. * scale_factor, path, fill_color)
> +        ).format(2.*scale_factor, path, fill_color)
>  
>      def contains(self, other):
>          """
> @@ -1580,10 +1632,9 @@ def contains(self, other):
>          >>> s2 = Segment3D(p2, p1)
>          >>> s.contains(s2)
>          True
> -        >>> s.contains((p1 + p2) / 2)
> +        >>> s.contains((p1 + p2)/2)
>          True
>          """
> -
>          if not isinstance(other, GeometryEntity):
>              other = Point(other, dim=self.ambient_dimension)
>          if isinstance(other, Point):
> @@ -1820,7 +1871,6 @@ def bounds(self):
>          rectangle for the geometric figure.
>  
>          """
> -
>          verts = self.points
>          xs = [p.x for p in verts]
>          ys = [p.y for p in verts]
> @@ -1985,7 +2035,6 @@ def _svg(self, scale_factor=1., fill_color="#66cc99"):
>          fill_color : str, optional
>              Hex string for fill color. Default is "#66cc99".
>          """
> -
>          from sympy.core.evalf import N
>  
>          verts = (N(self.p1), N(self.p2))
> @@ -1996,7 +2045,7 @@ def _svg(self, scale_factor=1., fill_color="#66cc99"):
>              '<path fill-rule="evenodd" fill="{2}" stroke="#555555" '
>              'stroke-width="{0}" opacity="0.6" d="{1}" '
>              'marker-start="url(#markerReverseArrow)" marker-end="url(#markerArrow)"/>'
> -            ).format(2. * scale_factor, path, fill_color)
> +        ).format(2.*scale_factor, path, fill_color)
>  
>      @property
>      def coefficients(self):
> @@ -2338,7 +2387,6 @@ def _svg(self, scale_factor=1., fill_color="#66cc99"):
>          fill_color : str, optional
>              Hex string for fill color. Default is "#66cc99".
>          """
> -
>          from sympy.core.evalf import N
>  
>          verts = (N(self.p1), N(self.p2))
> @@ -2347,7 +2395,7 @@ def _svg(self, scale_factor=1., fill_color="#66cc99"):
>          return (
>              '<path fill-rule="evenodd" fill="{2}" stroke="#555555" '
>              'stroke-width="{0}" opacity="0.6" d="{1}" />'
> -            ).format(2. * scale_factor, path, fill_color)
> +        ).format(2.*scale_factor, path, fill_color)
>  
>  
>  class LinearEntity3D(LinearEntity):
> @@ -2368,7 +2416,6 @@ class LinearEntity3D(LinearEntity):
>  
>      This is a base class and is not meant to be instantiated.
>      """
> -
>      def __new__(cls, p1, p2, **kwargs):
>          p1 = Point3D(p1, dim=3)
>          p2 = Point3D(p2, dim=3)
> @@ -2457,7 +2504,6 @@ class Line3D(LinearEntity3D, Line):
>      >>> L.points
>      (Point3D(2, 3, 4), Point3D(3, 5, 1))
>      """
> -
>      def __new__(cls, p1, pt=None, direction_ratio=[], **kwargs):
>          if isinstance(p1, LinearEntity3D):
>              if pt is not None:
> @@ -2509,9 +2555,9 @@ def equation(self, x='x', y='y', z='z', k=None):
>          """
>          if k is not None:
>              SymPyDeprecationWarning(
> -                            feature="equation() no longer needs 'k'",
> -                            issue=13742,
> -                            deprecated_since_version="1.2").warn()
> +                feature="equation() no longer needs 'k'",
> +                issue=13742,
> +                deprecated_since_version="1.2").warn()
>          from sympy import solve
>          x, y, z, k = [_symbol(i, real=True) for i in (x, y, z, 'k')]
>          p1, p2 = self.points
> @@ -2574,7 +2620,6 @@ class Ray3D(LinearEntity3D, Ray):
>      [1, 2, -4]
>  
>      """
> -
>      def __new__(cls, p1, pt=None, direction_ratio=[], **kwargs):
>          from sympy.utilities.misc import filldedent
>          if isinstance(p1, LinearEntity3D):
> @@ -2731,7 +2776,6 @@ class Segment3D(LinearEntity3D, Segment):
>      Point3D(5/2, 2, 8)
>  
>      """
> -
>      def __new__(cls, p1, p2, **kwargs):
>          p1 = Point(p1, dim=3)
>          p2 = Point(p2, dim=3)
> diff --git a/sympy/geometry/tests/test_line.py b/sympy/geometry/tests/test_line.py
> index 30f9a7e..b343565 100644
> --- a/sympy/geometry/tests/test_line.py
> +++ b/sympy/geometry/tests/test_line.py
> @@ -752,6 +752,20 @@ def test_parameter_value():
>      raises(ValueError, lambda: l.parameter_value((0, 0), t))
>  
>  
> +def test_bisectors():
> +    r1 = Line3D(Point3D(0, 0, 0), Point3D(1, 0, 0))
> +    r2 = Line3D(Point3D(0, 0, 0), Point3D(0, 1, 0))
> +    bisections = r1.bisectors(r2)
> +    assert bisections == [Line3D(Point3D(0, 0, 0), Point3D(1, 1, 0)),
> +        Line3D(Point3D(0, 0, 0), Point3D(1, -1, 0))]
> +    ans = [Line3D(Point3D(0, 0, 0), Point3D(1, 0, 1)),
> +        Line3D(Point3D(0, 0, 0), Point3D(-1, 0, 1))]
> +    l1 = (0, 0, 0), (0, 0, 1)
> +    l2 = (0, 0), (1, 0)
> +    for a, b in cartes((Line, Segment, Ray), repeat=2):
> +        assert a(*l1).bisectors(b(*l2)) == ans
> +
> +
>  def test_issue_8615():
>      a = Line3D(Point3D(6, 5, 0), Point3D(6, -6, 0))
>      b = Line3D(Point3D(6, -1, 19/10), Point3D(6, -1, 0))
> ```

Jaime02 commented:
Ups, I closed the PR accidentally. @schmir , I really appreciate your help. The diff file says that it is corrupted at line 319

smichr commented:
Make sure you have a blank nonindented last line in the diff. I needed to add this when I went to the diff page and copied and pasted it into a file. It applied without error to the current master. 

If you want, you can replace your master with my `linbis` branch, then `git reset head~1` and commit the results and push it here, too.

smichr commented:
OK, if you download my `linbis` branch and replace your master branch with it and push it here it will update everything. It has a single commit with you as the author.

Jaime02 commented:
Oh have you done it for me? Sorry for not answering, I'm busy.
Is eveything correct now?
